tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/mhlo_mark_shape_calc.cc

```diff
@@ -0,0 +1,313 @@
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/DenseSet.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Debug.h"
+#include "mlir-hlo/Dialect/mhlo/transforms/PassDetail.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"  // TF:llvm-project
+#include "mlir/IR/MLIRContext.h"              // TF:llvm-project
+#include "mlir/Pass/Pass.h"                   // TF:local_config_mlir
+#include "tensorflow/compiler/mlir/hlo/include/mlir-hlo/Dialect/mhlo/IR/hlo_ops.h"
+#include "tensorflow/compiler/mlir/hlo/include/mlir-hlo/utils/hlo_utils.h"
+
+namespace mlir {
+
+using hlo::kCpu;
+using hlo::kDiscShapeCalcAttr;
+using hlo::kGpu;
+using hlo::PlacementType;
+
+namespace mhlo {
+namespace {
+
+// Check Op if it is a mhlo Op.
+bool isMhloDialect(Operation* op) {
+  return (op->getDialect() ==
+          op->getContext()->getLoadedDialect<mhlo::MhloDialect>());
+}
+
+// This pass explicitly marks the shape calculating Op by adding an Attr. Nested
+// FuncOps should be taken into consideration.
+// Following Ops are shape Ops:
+//  - i64 Scalar output
+//  - Shape Op's operands
+//  - Shape operands according to kShapeCalcOperandMap
+// Following Ops regard as shape Ops:
+//  - GetDimensionSizeOp, PrintOp
+//  - ConstOp, SelectOp, IotaOp, DynamicIotaOp if type is i32
+//  - mhlo.dynamic_gather and mhlo.gather if operand_0's type is i32
+//  - Date operands but type is i32 according to kShapeCalcOperandMap
+struct MarkShapeCalc : public MarkShapeCalculationPassBase<MarkShapeCalc> {
+ public:
+  using MarkShapeCalculationPassBase<
+      MarkShapeCalc>::MarkShapeCalculationPassBase;
+
+  MarkShapeCalc() = default;
+  MarkShapeCalc(const MarkShapeCalc& o) {}
+
+  StringRef getArgument() const final { return "mhlo-disc-mark-shape-calc"; }
+
+  void runOnOperation() override;
+
+ private:
+  // Mark shape calculation subgraph
+  void MarkShapeCalcOps();
+
+  // Regard any mhlo Ops that calculates I32 as shape calculation Ops
+  void MarkRegardAsShapeCalcOps();
+
+  // for rule based placement strategy, the placement of the op in the list
+  // is up to the placement of the dominant operand
+  const DenseMap<StringRef, /*dominant operand index*/ int> kPlaceRuleMap = {
+      {"mhlo.dynamic_gather", /*operand*/ 0}, {"mhlo.gather", /*operand*/ 0}};
+
+  const DenseMap<StringRef, DenseSet<int>> kShapeCalcOperandMap = {
```

Comment:
I don't think we benefit from having a DenseSet instead of a `SmallVector<int, 3>` (or `std::vector`) here.
(see also below where it is used).

Reply:
Done.